### ISSUE TITLE
ZIO uses entropy sources directly and preventing us from writing determenistic async tests

### ISSUE THREAD

**Issue by pshirshov** [scalaz/scalaz-zio#310]
At the moment ZIO relies on external entropy sources which cannot be substituted by determenistic implementations:

```
  /** Determine the current system time **/
  val currentTime: IO[Nothing, Instant] = IO.sync(Instant.now)

  /** Determine the current system time in milliseconds **/
  val currentTimeMillis: IO[Nothing, Long] = IO.sync(System.currentTimeMillis)

  /** Returns the current value of the running Java Virtual Machine's high-resolution time source, in nanoseconds.
   * This method can only be used to measure elapsed time and is not related to any other notion of system or wall-clock time
   */
  val nanoTime: IO[Nothing, Long] = IO.sync(System.nanoTime)
```

`Schedule` uses these methods directly, so it's literally impossible to write a determenistic repeatable test for an asynchronous protocol involving `Schedule`.

Another problematic place is:

```
  final def jittered(min: Double, max: Double): Schedule[A, B] =
    modifyDelay((_, d) => IO.sync(util.Random.nextDouble()).map(random => d * min * (1 - random) + d * max * random))
```

I think all such things must be hidden behind an interface and it must be possible to customize them by overriding an `RTS` method.

**Comment by pshirshov** [comment]
By the way, an example of such an interface: 

```
trait Entropy {
  def nextBoolean(): Boolean

  def nextBytes(bytes: Array[Byte]): Unit

  def nextDouble(): Double

  def nextFloat(): Float

  def nextGaussian(): Double

  def nextLong(): Long

  def nextInt(max: Int): Int

  def nextInt(): Int

  def nextTimeUUID(): UUID

  def nextUUID(): UUID

  def shuffle[T, CC[X] <: TraversableOnce[X]](xs: CC[T])(implicit bf: CanBuildFrom[CC[T], T, CC[T]]): CC[T]
}
```

**Comment by jdegoes** [comment]
👍 This is related to #245, #309

**Comment by jdegoes** [comment]
Oh, and the original one: #140 

I know @Vilkina got busy with stuff and maybe someone can chip in with the remaining work.

**Comment by Vilkina** [comment]
@jdegoes will continue after 1st November.

**Comment by Vilkina** [comment]
kip going

**Comment by jdegoes** [comment]
The only thing remaining is #359

**Comment by pshirshov** [comment]
Thank you for the hard work on this. And I'm sorry that I was unable to pull myself out of business and work on this. Though hope you like @kaishh's contributions :)

**Comment by jdegoes** [comment]
Finally fixed by #359.

**Comment by jdegoes** [comment]
@pshirshov Yes, it's been great having @kaishh contributing!
